File: infra/devices.py
from common import config
from common.ids import device_id, fraud_device_id
from common.random import Rng
from common.timeline import sample_span, sample_short_span

from entities.models import People
from infra._pool import swap_delete
from infra.models import Devices, DeviceUsage, RingPlan

_DEVICE_TYPES: tuple[str, ...] = ("android", "ios", "web", "desktop")


def build(
    window_cfg: config.Window,
    patterns_cfg: config.Patterns,
    rng: Rng,
    people: People,
    ring_plans: dict[int, RingPlan],
) -> Devices:
    registry: dict[str, tuple[str, int]] = {}
    has_used: list[DeviceUsage] = []
    by_person: dict[str, list[str]] = {person_id: [] for person_id in people.ids}
    ring_map: dict[int, str] = {}

    start = window_cfg.start_date
    days = int(window_cfg.days)

    for person_id in people.ids:
        n_dev = 2 if rng.coin(0.20) else 1

        for idx in range(1, n_dev + 1):
            dev_id = device_id(person_id, idx)
            device_type = rng.choice(_DEVICE_TYPES)

            registry[dev_id] = (device_type, 0)
            by_person[person_id].append(dev_id)

            first_seen, last_seen = sample_span(rng, start, days)
            has_used.append(
                DeviceUsage(
                    person_id=person_id,
                    device_id=dev_id,
                    first_seen=first_seen,
                    last_seen=last_seen,
                )
            )

    legit_pool = [
        person_id for person_id in people.ids if person_id not in people.frauds
    ]
    remaining_legit = list(legit_pool)
    remaining_index = {person_id: idx for idx, person_id in enumerate(remaining_legit)}

    legit_shared_counter = 1
    noise_p = float(patterns_cfg.legit_device_noise_p)

    for anchor in legit_pool:
        if anchor not in remaining_index:
            continue

        if not rng.coin(noise_p):
            continue

        swap_delete(remaining_legit, remaining_index, anchor)

        if not remaining_legit:
            break

        extra_count = min(len(remaining_legit), rng.int(1, 3))
        peers = rng.choice_k(remaining_legit, extra_count, replace=False)
        group = [anchor, *peers]

        shared_id = f"LD{legit_shared_counter:06d}"
        legit_shared_counter += 1

        registry[shared_id] = (rng.choice(_DEVICE_TYPES), 0)
        first_seen, last_seen = sample_short_span(rng, start, days)

        for person_id in group:
            by_person[person_id].append(shared_id)
            has_used.append(
                DeviceUsage(
                    person_id=person_id,
                    device_id=shared_id,
                    first_seen=first_seen,
                    last_seen=last_seen,
                )
            )

        for person_id in peers:
            if person_id in remaining_index:
                swap_delete(remaining_legit, remaining_index, person_id)

    for plan in ring_plans.values():
        shared_id = fraud_device_id(plan.ring_id)
        registry[shared_id] = (rng.choice(_DEVICE_TYPES), 1)
        ring_map[plan.ring_id] = shared_id

        for person_id in plan.shared_device_members:
            by_person[person_id].append(shared_id)
            has_used.append(
                DeviceUsage(
                    person_id=person_id,
                    device_id=shared_id,
                    first_seen=plan.first_seen,
                    last_seen=plan.last_seen,
                )
            )

    return Devices(
        registry=registry,
        has_used=has_used,
        by_person=by_person,
        ring_map=ring_map,
    )

File: src/synth/entities/infra/devices.cpp
#include "phantomledger/synth/entities/infra/devices.hpp"

#include "phantomledger/synth/entities/infra/pool.hpp"
#include "phantomledger/synth/entities/infra/timeline.hpp"

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <unordered_map>
#include <utility>
#include <vector>

namespace PhantomLedger::infra::synth::devices {

using DeviceIdentity = ::PhantomLedger::devices::Identity;

namespace {

[[nodiscard]] DeviceKind sampleKind(random::Rng &rng) {
  const auto idx = rng.choiceIndex(kAllDeviceKinds.size());
  return kAllDeviceKinds[idx];
}

void registerRecord(Output &out, DeviceIdentity id, DeviceKind kind,
                    bool flagged) {
  out.records.push_back(Record{
      .identity = id,
      .kind = kind,
      .flagged = flagged,
  });
}

[[nodiscard]] std::vector<entity::PersonId>
buildLegitPool(const entity::person::Roster &people) {
  std::vector<entity::PersonId> out;
  out.reserve(people.count);
  for (entity::PersonId p = 1; p <= people.count; ++p) {
    if (!people.has(p, entity::person::Flag::fraud)) {
      out.push_back(p);
    }
  }
  return out;
}

} // namespace

Output AssignmentRules::build(
    random::Rng &rng, time::Window window, const entity::person::Roster &people,
    const std::unordered_map<std::uint32_t, RingPlan> &ringPlans) const {
  Output out;
  const auto reserveHint =
      static_cast<std::size_t>(people.count) * 12U / 10U + ringPlans.size();
  out.records.reserve(reserveHint);
  out.usages.reserve(reserveHint);

  for (entity::PersonId p = 1; p <= people.count; ++p) {
    out.byPerson.try_emplace(p);
  }

  const auto windowStart = window.start;
  const auto windowDays = window.days;

  for (entity::PersonId p = 1; p <= people.count; ++p) {
    const std::uint32_t nDev = rng.coin(secondDeviceP) ? 2U : 1U;

    for (std::uint32_t slot = 1; slot <= nDev; ++slot) {
      const auto id = DeviceIdentity::person(p, slot);
      const auto kind = sampleKind(rng);

      registerRecord(out, id, kind, /*flagged=*/false);
      out.byPerson[p].push_back(id);

      const auto [firstSeen, lastSeen] =
          timeline::sampleSpan(rng, windowStart, windowDays);
      out.usages.push_back(Usage{
          .personId = p,
          .deviceId = id,
          .firstSeen = firstSeen,
          .lastSeen = lastSeen,
      });
    }
  }

  const auto legitPool = buildLegitPool(people);

  std::vector<entity::PersonId> remaining = legitPool;
  std::unordered_map<entity::PersonId, std::size_t> remainingIndex;
  remainingIndex.reserve(remaining.size());
  for (std::size_t i = 0; i < remaining.size(); ++i) {
    remainingIndex[remaining[i]] = i;
  }

  std::uint64_t legitSharedCounter = 1;

  for (const auto anchor : legitPool) {
    if (!pool::contains(remainingIndex, anchor)) {
      continue;
    }
    if (!rng.coin(legitDeviceNoiseP)) {
      continue;
    }

    pool::swapDelete(remaining, remainingIndex, anchor);

    if (remaining.empty()) {
      break;
    }

    const auto cap = std::min<std::size_t>(remaining.size(), 3U);
    const auto extraCount = static_cast<std::size_t>(
        rng.uniformInt(1, static_cast<std::int64_t>(cap) + 1));

    const auto pickIdx = rng.choiceIndices(remaining.size(), extraCount, false);
    std::vector<entity::PersonId> peers;
    peers.reserve(extraCount);
    for (const auto i : pickIdx) {
      peers.push_back(remaining[i]);
    }

    const auto sharedId =
        ::PhantomLedger::devices::legitShared(legitSharedCounter, 0);
    ++legitSharedCounter;

    const auto kind = sampleKind(rng);
    registerRecord(out, sharedId, kind, /*flagged=*/false);

    const auto [firstSeen, lastSeen] =
        timeline::sampleShortSpan(rng, windowStart, windowDays);

    // The anchor is part of the group too.
    std::vector<entity::PersonId> group;
    group.reserve(peers.size() + 1);
    group.push_back(anchor);
    for (const auto pid : peers) {
      group.push_back(pid);
    }

    for (const auto pid : group) {
      out.byPerson[pid].push_back(sharedId);
      out.usages.push_back(Usage{
          .personId = pid,
          .deviceId = sharedId,
          .firstSeen = firstSeen,
          .lastSeen = lastSeen,
      });
    }

    for (const auto pid : peers) {
      if (pool::contains(remainingIndex, pid)) {
        pool::swapDelete(remaining, remainingIndex, pid);
      }
    }
  }

  std::vector<std::uint32_t> ringIds;
  ringIds.reserve(ringPlans.size());
  for (const auto &kv : ringPlans) {
    ringIds.push_back(kv.first);
  }
  std::sort(ringIds.begin(), ringIds.end());

  for (const auto ringId : ringIds) {
    const auto &plan = ringPlans.at(ringId);

    const auto sharedId = DeviceIdentity::ring(plan.ringId, 0);
    const auto kind = sampleKind(rng);
    registerRecord(out, sharedId, kind, /*flagged=*/true);
    out.ringMap.emplace(plan.ringId, sharedId);

    for (const auto pid : plan.sharedDeviceMembers) {
      out.byPerson[pid].push_back(sharedId);
      out.usages.push_back(Usage{
          .personId = pid,
          .deviceId = sharedId,
          .firstSeen = plan.firstSeen,
          .lastSeen = plan.lastSeen,
      });
    }
  }

  return out;
}

} // namespace PhantomLedger::infra::synth::devices
